pkg/controller/direct/workstations/workstation_controller.go

```diff
@@ -0,0 +1,259 @@
+func (a *WorkstationAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("updating Workstation", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	desiredPb := WorkstationSpec_ToProto(mapCtx, &a.desired.DeepCopy().Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	// Set name and etag manually, because they are not filled-in by WorkstationConfigSpec_ToProto.
+	desiredPb.Name = a.id.String()
+	desiredPb.Etag = a.actual.Etag
+
+	paths, err := common.CompareProtoMessage(desiredPb, a.actual, common.BasicDiff)
+	if err != nil {
+		return err
+	}
+	if len(paths) == 0 {
+		log.V(2).Info("no field needs update", "name", a.id.String())
+		status := &krm.WorkstationStatus{}
+		status.ObservedState = WorkstationObservedState_FromProto(mapCtx, a.actual)
+		if mapCtx.Err() != nil {
+			return mapCtx.Err()
+		}
+		return updateOp.UpdateStatus(ctx, status, nil)
+	}
+	updateMask := &fieldmaskpb.FieldMask{Paths: sets.List(paths)}
+
+	req := &workstationspb.UpdateWorkstationRequest{
+		Workstation: desiredPb,
+		UpdateMask:  updateMask,
+	}
+	op, err := a.gcpClient.UpdateWorkstation(ctx, req)
+	if err != nil {
+		return fmt.Errorf("updating Workstation %s: %w", a.id.String(), err)
+	}
+	updated, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("waiting Workstation %s update: %w", a.id.String(), err)
+	}
+	log.V(2).Info("successfully updated Workstation", "name", a.id.String())
+
+	status := &krm.WorkstationStatus{}
+	status.ObservedState = WorkstationObservedState_FromProto(mapCtx, updated)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	return updateOp.UpdateStatus(ctx, status, nil)
+}
+
+// Export maps the GCP object to a Config Connector resource `spec`.
+func (a *WorkstationAdapter) Export(ctx context.Context) (*unstructured.Unstructured, error) {
+	if a.actual == nil {
+		return nil, fmt.Errorf("Find() not called")
+	}
+	u := &unstructured.Unstructured{}
+
+	obj := &krm.Workstation{}
+	mapCtx := &direct.MapContext{}
+	obj.Spec = direct.ValueOf(WorkstationSpec_FromProto(mapCtx, a.actual))
+	if mapCtx.Err() != nil {
+		return nil, mapCtx.Err()
+	}
+	obj.Spec.Parent = &krm.WorkstationConfigRef{External: a.id.Parent().String()}
+	uObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	u.SetName(a.id.ID())
+	u.SetGroupVersionKind(krm.WorkstationGVK)
+
+	u.Object = uObj
+	return u, nil
+}
+
+// Delete the resource from GCP service when the corresponding Config Connector resource is deleted.
+func (a *WorkstationAdapter) Delete(ctx context.Context, deleteOp *directbase.DeleteOperation) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("deleting Workstation", "name", a.id)
+
+	req := &workstationspb.DeleteWorkstationRequest{Name: a.id.String()}
+	op, err := a.gcpClient.DeleteWorkstation(ctx, req)
+	if err != nil {
```

Comment:
Does the GCP service returns any code if the resource does not exist (like 404)? If so, we may want to exclude NotFound from the error. If this is included in the current code, would you mind adding a comment?

Reply:
Nice catch. Fixed. Also noticed that WorkstationConfig and WorkstationCluster have this same issue, so I fixed those as well.